pkg/sql/sqlstats/persistedsqlstats/provider.go

```diff
@@ -194,8 +194,16 @@ func (s *PersistedSQLStats) startSQLStatsFlushLoop(ctx context.Context, stopper
 				return
 			}
 
+			// Flush may do some extra work if the SQLStatsFlushEnabled is set to false,
+			// so we don't want to exit earlier than necessary.
+			// TODO (xinhaoz): Investigate if we can lift that step out of the flush function,
+			// so we can skip here.
 			s.Flush(ctx, stopper)
 
+			if !SQLStatsFlushEnabled.Get(&s.cfg.Settings.SV) {
```

Comment:
I generally want to avoid checking a bool in multiple parts of the codebase because then it's hard to reason about the consequences of flipping the setting.

If you think avoiding the work below is important, what do you think about `s.Flush` returning a bool telling us whether it did anything. Then we can avoid signaling when it returns `false`. Then it keeps the code chunks in sync even if the setting changes in the middle and avoids us having to think about the possibility.

Also, if we're adding conditional logic to the job we should add a test to exercise the change.

Reply:
Good idea on the bool - done and test added. We can consider renaming `Flush` to `MaybeFlush` or something if desired.